go/vt/mysqlctl/backupstats/stats.go

```diff
@@ -0,0 +1,199 @@
+/*
+Copyright 2022 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package backupstats
+
+import (
+	"sync"
+	"time"
+
+	"vitess.io/vitess/go/stats"
+	vtstats "vitess.io/vitess/go/stats"
+)
+
+// Stats is a reporting interface meant to be shared amount backup and restore
+// components.
+//
+// This interface is meant to give those components a way to reports stats
+// without having to register their own stats or to import globally registered
+// stats. This way out-of-tree plugins can have a mechanism for reporting
+// stats, while the policies for those stats (metric names and labels, stat
+// sinks) remain in the control of in-tree Vitess code and Vitess users.
+type Stats interface {
+	// Scope creates a new Stats which inherits this Stats' scopes plus any
+	// new provided scopes.
+	//
+	// Once a ScopeType has been set in a Stats, it cannot be changed by
+	// further calls to Scope().
+	//
+	// This allows parent components to prepare properly scoped Stats and pass
+	// them to child components in such a way that child components cannot
+	// overwrite another components' metrics.
+	Scope(...Scope) Stats
+	// Increment count by 1 and increase duration.
+	TimedIncrement(time.Duration)
+	// Increment bytes and increase duration.
+	TimedIncrementBytes(int, time.Duration)
+}
+
+type noStats struct{}
+
+type scopedStats struct {
+	bytes       *vtstats.CountersWithMultiLabels
+	count       *vtstats.CountersWithMultiLabels
+	durationNs  *vtstats.CountersWithMultiLabels
+	labelValues []string
+}
+
+const unscoped = "-"
+
+var (
+	defaultNoStats *noStats
+
+	labels = []string{"component", "implementation", "operation"}
+
+	registerBackupStats  sync.Once
+	registerRestoreStats sync.Once
+
+	backupBytes       *stats.CountersWithMultiLabels
+	backupCount       *stats.CountersWithMultiLabels
+	backupDurationNs  *stats.CountersWithMultiLabels
+	restoreBytes      *stats.CountersWithMultiLabels
+	restoreCount      *stats.CountersWithMultiLabels
+	restoreDurationNs *stats.CountersWithMultiLabels
+)
+
+// BackupStats creates a new Stats for backup operations.
+//
+// It registers two stats with the Vitess stats package.
```

Comment:
```suggestion
// It registers three stats with the Vitess stats package.
```

Reply:
Or just say "the following stats", then the comment won't become incorrect when we add the next metric.

Reply:
stats vs metrics? Metrics is probably more correct.